extensions/rotating-teleport/common/client.mjs

```diff
@@ -0,0 +1,223 @@
+ */
+class Rotator {
+  /**
+   * @param {Number} zeroAngle - unit: deg
+   * @param {Number} time - ROTATE_TIME_FORWARD or ROTATE_TIME_BACKWARD
+   */
+  constructor(zeroAngle, time=ROTATE_TIME_FORWARD) {
+    this.t0 = null;
+    this.x0 = zeroAngle;
+    this.forwardTime = (time === ROTATE_TIME_FORWARD);
+  }
+
+  start() {
+    this.t0 = Date.now() / 1000;
+  }
+
+  getStatus() {
+    const dt = Date.now() / 1000 - this.t0;
+    const dt2 = (this.forwardTime) ?
+        dt * dt :
+        (ROTATE_HALF_DURATION - dt) * (ROTATE_HALF_DURATION - dt);
+    return {
+      end: dt >= ROTATE_HALF_DURATION,
+      angle: this.x0 + ROTATE_ANGULAR_ACCELERATION * dt2,
+    };
+  }
+
+  static angle2direction(angle) {
+    const normalizedAngle = (0 <= angle && angle < 360) ? angle : ((angle % 360 + 360) % 360);
+    if (45 <= normalizedAngle && normalizedAngle < 135) return 'U';
+    if (135 <= normalizedAngle && normalizedAngle < 225) return 'L';
+    if (225 <= normalizedAngle && normalizedAngle < 315) return 'D';
+    return 'R';
+  }
+
+  static direction2angle(direction) {
+    switch (direction) {
+      case 'U':
+        return 90;
+      case 'L':
+        return 180;
+      case 'D':
+        return 270;
+      default:
+        return 0;
+    }
+  }
+}
+
+
+/**
+ * TODO: jsdoc
+ */
+class Client {
+  /**
+   * Create the client side of the extension.
+   * @constructor
+   * @param {ClientExtensionHelper} helper - An extension helper object for
+   * servicing various functionalities of the extension.
+   */
+  constructor(helper) {
+    this.helper = helper;
+    this.teleportLookupTable = new Map(); // key: the ID of the portal, value: the destination
+
+    // If the player is being teleported, then there would be an entry in this map.
+    // key: playerID, value: the target place of teleportation
+    this.isBeingTeleported = new Map();
+  }
+
+  /**
+   * The initialization function.
+   */
+  async gameStart() {
+    this.parseLayer();
+
+    // teleport the player if he/she steps on a portal
+    this.helper.gameState.registerOnPlayerUpdate((msg) => {
+      if (typeof msg.playerID !== 'string') return;
+      if (typeof msg.mapCoord !== 'object') return;
+
+      // if he/she is not being teleported, check if steps on a portal defined by our extension
+      const layerValue = this.helper.gameMap.getCell(msg.mapCoord, LAYER_NAME);
+      if (layerValue === null) return;
+      const [id, action] = layerValue.split(SEPARATOR);
+      if (action !== ACTION_IN) return;
+
+      // rotating teleport
+      const fromPlace = this.helper.gameState.getPlayer(msg.playerID).mapCoord.copy();
+      const targetPlace = this.teleportLookupTable.get(id);
+      this.tryStartRotatingTeleportAnimation(msg.playerID, fromPlace, targetPlace);
+    });
+  }
+
+  /**
+   * Parse the configuration
+   * @return {Boolean} - successful or not
+   */
+  parseLayer() {
+    // parse the layers and build the lookup table
+    const inIDs = new Set();
```

Comment:
Optional: Add a comment that we use map-portal mostly.

Reply:
I think the structure of map-portal is not friendly to development (using a pre-defined layer instead of using post-processing with cellset). I would like to see it refactored in the future, so I won't add the comment here.